### PR TITLE
feat: Agilent-Gen5 Multiple Read Modes

### PR DESCRIPTION
**Pull request by sumaga-benchling**
Added multi-read mode functionality to Agilent Gen5 adapter with reference to tickets **CUST-8062** and **BNCH-102635** on JIRA. Ready to be merged after review.

### CODE REVIEW

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 1]
```diff
@@ -268,42 +268,65 @@ def create(cls, reader: LinesReader) -> ReadData:
...
+        read_modes = cls.get_read_modes(procedure_details)
+        read_data_list: [ReadData] = []
+        full_read_lines: list[str] = procedure_details.splitlines()
+
+        section_lines_reader = SectionLinesReader(full_read_lines)
+        read_sections = []
+        for read_section in section_lines_reader.iter_sections("^Read\t"):
+            read_sections.append(read_section)
+
+        section = 0
```
Comment (nathan-stender) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:280]: This index is not used except to access read_sections, please use:

```
for section in read_sections:
```

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 2]
```diff
@@ -268,42 +268,65 @@ def create(cls, reader: LinesReader) -> ReadData:
...
+        for read_mode_idx in range(len(read_modes)):
```
Comment (nathan-stender) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:281]: Similarly, this is no used except to access read_modes, use:

```
for read_mode in read_modes:
```

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 3]
```diff
@@ -268,42 +268,65 @@ def create(cls, reader: LinesReader) -> ReadData:
...
+
         if not read_modes:
             raise AllotropeConversionError(UNSUPPORTED_READ_MODE_ERROR)
-        if len(read_modes) > 1:
-            raise AllotropeConversionError(MULTIPLE_READ_MODE_ERROR)
 
-        if read_modes[0] in (ReadMode.FLUORESCENCE, ReadMode.ALPHALISA):
-            return ReadMode.FLUORESCENCE
-        return read_modes[0]
+        # Replace ALPHALISA with FLUORESCENCE
```
Comment (nathan-stender) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:323]: Handle case where there is `ALPHALISA` and `FLUORESCENCE`. If we expect it to never happen, check for it and raise an `AllotropeConversionError`

File: src/allotropy/allotrope/schema_mappers/adm/plate_reader/benchling/_2023/_09/plate_reader.py [hunk 4]
```diff
@@ -64,10 +65,10 @@ class MeasurementType(Enum):
 
 
 MeasurementDocumentItems = (
-    UltravioletAbsorbancePointDetectionMeasurementDocumentItems
-    | FluorescencePointDetectionMeasurementDocumentItems
-    | LuminescencePointDetectionMeasurementDocumentItems
-    | OpticalImagingMeasurementDocumentItems
+        UltravioletAbsorbancePointDetectionMeasurementDocumentItems
```
Comment (nathan-stender) [inline src/allotropy/allotrope/schema_mappers/adm/plate_reader/benchling/_2023/_09/plate_reader.py:68]: You seem to have some kind of odd setting that's causing you to add an extra indent when formating. Please revert these changes. 

It might be easiest to find the setting and update it, and then run hatch run fix

Unfortunately I've never seen this kind of change, so I'm not sure what the setting is.

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 5]
```diff
@@ -540,10 +570,10 @@ def _get_sources(
     return sources or measurements
 
 
-def _create_metadata(header_data: HeaderData, read_data: ReadData) -> Metadata:
+def _create_metadata(header_data: HeaderData, read_data: list[ReadData]) -> Metadata:
     return Metadata(
         device_type=DEVICE_TYPE,
-        detection_type=read_data.read_mode.value,
+        detection_type=read_data[0].read_mode.value,
```
Comment (nathan-stender) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:576]: This doesn't make sense to set here, since the detection type does not apply to all measurements.

You handle this in the schema mapper by hard-coding the read mode, which is fine, but it means this is not used.

Please remove this, and make `detection_type` optional in the `Metadata` class

File: tests/parsers/agilent_gen5/agilent_gen5_structure_test.py [hunk 6]
```diff
@@ -17,18 +17,18 @@
 @pytest.mark.short
 def test_create_header_data_no_well_plate_id_in_filename() -> None:
     header_rows = [
-        "Software Version	3.12.08",
+        "Software Version   3.12.08",
```
Comment (nathan-stender) [inline tests/parsers/agilent_gen5/agilent_gen5_structure_test.py:20]: Similarly - I don't know what's happening with formatting here, but please fix before we merge

File: tests/parsers/agilent_gen5/agilent_gen5_structure_test.py [hunk 7]
```diff
@@ -110,12 +110,12 @@ def test_create_read_data_absorbance() -> None:
 
     read_data = ReadData.create(reader)
 
-    assert read_data.read_mode == ReadMode.ABSORBANCE
-    assert read_data.pathlength_correction == "977 / 900"
-    assert read_data.step_label == "260"
-    assert read_data.detector_carriage_speed == "Normal"  # Read Speed
-    assert read_data.number_of_averages == 8  # Measurements/Data Point
-    assert read_data.measurement_labels == [
+    assert read_data[0].read_mode == ReadMode.ABSORBANCE
```
Comment (nathan-stender) [inline tests/parsers/agilent_gen5/agilent_gen5_structure_test.py:113]: Add `assert len(read_data) == 1`

File: tests/parsers/agilent_gen5/agilent_gen5_structure_test.py [hunk 8]
```diff
@@ -143,12 +143,12 @@ def test_create_read_data_luminescence_full_light() -> None:
 
     read_data = ReadData.create(reader)
 
-    assert read_data.read_mode == ReadMode.LUMINESCENCE
-    assert read_data.step_label == "LUM"
-    assert read_data.detector_carriage_speed == "Normal"  # Read Speed
-    assert read_data.detector_distance == 4.5  # Read Height
-    assert read_data.measurement_labels == ["LUM:Lum"]
-    assert read_data.filter_sets == {
+    assert read_data[0].read_mode == ReadMode.LUMINESCENCE
```
Comment (nathan-stender) [inline tests/parsers/agilent_gen5/agilent_gen5_structure_test.py:146]: Ditto

File: tests/parsers/agilent_gen5/agilent_gen5_structure_test.py [hunk 9]
```diff
@@ -376,3 +376,160 @@ def test_create_layout_data_with_name_rows_name_row_first() -> None:
...
+        "\tFilter Set 1",
+        "\t\tExcitation: 579,  Emission: 616",
+        "\t\tOptics: Top,  Gain: extended",
+        "\tFilter Set 2",
+        "\t\tExcitation: 479,  Emission: 520",
+        "\t\tOptics: Top,  Gain: extended",
+    ]
+    reader = LinesReader(multiple_read_modes)
+
+    file_data = ReadData.create(reader)
```
Comment (nathan-stender) [inline tests/parsers/agilent_gen5/agilent_gen5_structure_test.py:403]: Nit: rename to `read_data`

File: tests/parsers/agilent_gen5/agilent_gen5_structure_test.py [hunk 10]
```diff
@@ -376,3 +376,160 @@ def test_create_layout_data_with_name_rows_name_row_first() -> None:
         "B2": "Name_B2",
         "B3": "Name_B3",
     }
+
+
+@pytest.mark.short
+def test_create_multiple_read_modes() -> None:
```
Comment (nathan-stender) [inline tests/parsers/agilent_gen5/agilent_gen5_structure_test.py:382]: Good job with lots of tests covering cases 👍

File: tests/parsers/agilent_gen5/testdata/multi_read_modes/jira_cust_8062_bug.txt [hunk 11]
```diff
@@ -0,0 +1,80 @@
+
```
Comment (nathan-stender) [inline tests/parsers/agilent_gen5/testdata/multi_read_modes/jira_cust_8062_bug.txt:1]: Can you please rename this from `jira_cust_8062_bug` to something like `absorbance_and_fluorescence`? We don't want to reference JIRA tickets in allotropy.

File: tests/parsers/agilent_gen5/agilent_gen5_structure_test.py [hunk 12]
```diff
@@ -376,3 +376,160 @@ def test_create_layout_data_with_name_rows_name_row_first() -> None:
...
+            emission="460/40",
+            gain="136",
+        )
+    }
+    assert file_data[2].detector_carriage_speed == "Normal"
+    assert file_data[2].detector_distance == 4.5
+
+
+@pytest.mark.short
+def test_create_two_same_read_modes() -> None:
```
Comment (nathan-stender) [inline tests/parsers/agilent_gen5/agilent_gen5_structure_test.py:498]: Can we please add a file test case for this as well, I would like to see the full code for adding measurements exercised here.

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 13]
```diff
@@ -268,42 +268,65 @@ def create(cls, reader: LinesReader) -> ReadData:
...
+
+        section_lines_reader = SectionLinesReader(full_read_lines)
+        read_sections = []
+        for read_section in section_lines_reader.iter_sections("^Read\t"):
+            read_sections.append(read_section)
+
+        section = 0
+        for read_mode_idx in range(len(read_modes)):
+            read_mode = read_modes[read_mode_idx]
+            while section < len(read_sections):
```
Comment (joshua-benchling) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:283]: you dont need a loop here since it´s only being iterated once. One thing you can do is create a dict where its key is the read mode and the value is the red sections lines. This way you only need the initial loop on the read modes.

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py (hunk 5 above)
Comment (joshua-benchling) [reply src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:576]: Agree, this field should be removed since we support multi read mode and we take it from the measurement.

**Review by joshua-benchling — COMMENTED** [review]
Please also check the contributions file on how to name the PR and if you need to add a change log.

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py (hunk 13 above)
Comment (nathan-stender) [reply src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py]: Good suggestion - we ended up zipping the two lists together

**Review by nathan-stender — APPROVED** [review]

**Comment by nathan-stender** [comment]
@rayaankha-tara I approve this PR, but I paired with @sumaga-benchling to deal with some issues, so I have commits and can't approve myself :D

File: tests/parsers/unchained_labs_lunatic/testdata/2024-07-12_CW288_Plate1.json [hunk 14]
```diff
@@ -16,6 +16,7 @@
                                 "device control document": [
                                     {
                                         "device type": "plate reader",
+                                        "detection type": "Absorbance",
```
Comment (nathan-stender) [inline tests/parsers/unchained_labs_lunatic/testdata/2024-07-12_CW288_Plate1.json:19]: NOTE to reviewer: these changes are because unchained parser previously wasn't adding detection type (because it was implied) but no harm in adding it, for consistency.

**Review by rayaankha-tara — APPROVED** [review]

**Review by joshua-benchling — APPROVED** [review]
